Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,443 @@
+
+- *Reconcilers*: control-loops that observe the desired state
+  tables and reconcile them against a target such as a BPF map or the Linux routing
+  table. The reconciler is usually an instance of the StateDB reconciler which is
+  defined in terms of a table of objects with a status field and the operations
+  *Update*, *Delete* and *Prune*.
+
+Dividing the agent this way we achieve a nice separation of concerns:
+
+* Separating the user intent into its own tables keeps the parsing and validation from the computation we'll perform on the data. It also makes it nicer to reuse as it's purely about representing the outside intent internally in an efficient way without tying it too much into implementation details of a specific feature.
+
+* By defining the controller as essentially the function from input tables to output tables it becomes easy to understand and test.
+
+* Separating the reconciliation from the desired state computation the complex logic of dealing with low-level errors and retrying is separate from the pure "business logic" computation.
+
+* Using the generic reconcilers allows using tried-and-tested and instrumented retry implementation.
+
+* The control-plane of the agent is essentially everything outside the reconcilers This allows us to integration test, simulate or benchmark the control-plane code without unreasonable amount of scaffolding. The easier it is to write reliable integration tests the more resilient the codebase becomes.
+
+What we're trying to achieve is well summarized by Fred Brooks in "The Mythical Man Month":
+
+	| Show me your flowchart and conceal your tables, and I shall continue to be mystified.
+	| Show me your tables, and I won't usually need your flowchart; it'll be obvious.
+
+Defining tables
+~~~~~~~~~~~~~~~
+
+`StateDB documentation <https://github.com/cilium/statedb/blob/main/README.md>`__ gives a good introduction into how to create a table and its indexes, so we won't repeat that here, but instead focus on Cilium specific details.
+
+Let's start off with some guidelines that you might want to consider:
+
+* By default publicly provide ``Table[Obj]`` so new features can build on it and it can be used in tests. Also export the table's indexes or the query functions (``var ByName = nameIndex.Query``).
+
+* Do not export ``RWTable[Obj]`` if outside modules do not need to directly write into the table. If other modules do write into the table, consider defining "writer functions" that validate that the writes are well-formed.
+
+* If the table is closely associated with a specific feature, define it alongside the implementation of the feature. If the table is shared by many modules, consider defining it in ``daemon/k8s`` or ``pkg/datapath/tables`` so it is easy to discover.
+
+* Make sure the object can be JSON marshalled so it can be inspected. If you need to store non-marshallable data (e.g. functions), make them private or mark them with ``json:"-"`` struct tag.
```

Comment:
should private fields be printed to the CLI and have a placeholder or just be ignored?

Reply:
Just ignored since they can't be transferred. If one needs to show information in a field that's not marshallable then you need a separate field to summarize that info for json marshalling (e.g. for `cilium-dbg statedb` commands and dumping).